In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn import metrics


In [ ]:
train_data = pd.read_csv("/content/drive/MyDrive/digit-recognizer/train.csv")
test_data = pd.read_csv("/content/drive/MyDrive/digit-recognizer/test.csv")

In [ ]:
train_data.isnull().sum()

label       0
pixel0      0
pixel1      0
pixel2      0
pixel3      0
           ..
pixel779    0
pixel780    0
pixel781    0
pixel782    0
pixel783    0
Length: 785, dtype: int64

In [ ]:
test_data.isnull().sum()

pixel0      0
pixel1      0
pixel2      0
pixel3      0
pixel4      0
           ..
pixel779    0
pixel780    0
pixel781    0
pixel782    0
pixel783    0
Length: 784, dtype: int64

In [ ]:
test_data.head()

pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 784 columns]

In [ ]:
train_data.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [ ]:
y = train_data['label']

In [ ]:
X = train_data.drop(columns = 'label')

In [ ]:
X = X/255.0
test_data = test_data/255.0

In [ ]:

from sklearn.preprocessing import scale
X_scaled = scale(X)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.3, train_size = 0.2 ,random_state = 10)

In [ ]:
#linear model
model_linear = SVC(kernel='linear')
model_linear.fit(X_train, y_train)
y_pred = model_linear.predict(X_test)

In [ ]:
print("accuracy:", metrics.accuracy_score(y_true=y_test, y_pred=y_pred), "\n")

accuracy: 0.9133333333333333 



In [ ]:
#non-linear model
non_linear_model = SVC(kernel='rbf')
non_linear_model.fit(X_train, y_train)
y_pred = non_linear_model.predict(X_test)

In [ ]:
print("accuracy:", metrics.accuracy_score(y_true=y_test, y_pred=y_pred), "\n")

accuracy: 0.9348412698412698 



In [ ]:
model = SVC(C=10, gamma=0.001, kernel="rbf")

model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print("accuracy", metrics.accuracy_score(y_test, y_pred), "\n")

accuracy 0.9438888888888889 



In [ ]:
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
print("Root Mean Sqaured Error:", rmse)

Root Mean Sqaured Error: 0.9641593097632645


In [ ]:
from sklearn.metrics import precision_score
precision = precision_score(y_test, y_pred,average='weighted')
print(f'Precision: {precision}')

Precision: 0.9445846687242071


In [ ]:
from sklearn.metrics import recall_score
recall = recall_score(y_test, y_pred,average='weighted')
print(f'Recall: {recall}')

Recall: 0.9438888888888889


In [ ]:
from sklearn.metrics import f1_score
f1 = f1_score(y_test, y_pred,average = 'weighted')
print("F1 Score:", f1)

F1 Score: 0.9439371998582876


In [ ]:
y_test

27084    7
18640    3
41477    9
39744    8
28354    6
        ..
36651    7
23060    9
10399    1
10740    0
25674    1
Name: label, Length: 12600, dtype: int64

In [ ]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[1163    0    4    1    1    2    8    6    3    0]
 [   0 1389    4    2    4    0    1    9    4    0]
 [   1    4 1184   14    5    1    9   30    7    5]
 [   0    3   15 1263    0   14    2   23    8    3]
 [   1    2   20    3 1149    0   10   10    2   21]
 [   2    8    3   30    4 1064   15    9   11    3]
 [   8    1    3    0    3   13 1167   23    1    0]
 [   4    9   10    8   12    0    0 1255    2   30]
 [   5   18   17   23    8   20    5   13 1098   10]
 [   5    3    2   27   21    1    1   51    3 1161]]


Accuracy: 94.38%
Precision: 94.46%
Recall: 94.38%
F1_score: 94.39%
Training Time: Faster than CNN's
Hyperparameters:
C (Regularization parameter): Controls the trade-off between achieving a low error on training data and minimizing the complexity of the model. 10
Gamma (Kernel coefficient): Defines how far the influence of a single training example reaches, impacting the decision boundary shape in the feature space. 0.001
Special features : handling non-linear classification tasks without explicitly transforming the input data. generally robust to noise in the data and outliers



In [27]:
from sklearn.metrics import classification_report
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.98      0.98      0.98      1188
           1       0.97      0.98      0.97      1413
           2       0.94      0.94      0.94      1260
           3       0.92      0.95      0.93      1331
           4       0.95      0.94      0.95      1218
           5       0.95      0.93      0.94      1149
           6       0.96      0.96      0.96      1219
           7       0.88      0.94      0.91      1330
           8       0.96      0.90      0.93      1217
           9       0.94      0.91      0.93      1275

    accuracy                           0.94     12600
   macro avg       0.95      0.94      0.94     12600
weighted avg       0.94      0.94      0.94     12600

